libyul/optimiser/SSATransform.cpp

```diff
@@ -130,7 +128,142 @@ void IntroduceSSA::operator()(Block& _block)
 }
 
 /**
- * Second step of SSA transform: Replace the references to variables-to-be-replaced
+ * Second step of SSA transform: Introduces new SSA variables at each control-flow join
+ * and at the beginning of functions.
+ */
+class IntroduceControlFlowSSA: public ASTModifier
+{
+public:
+	explicit IntroduceControlFlowSSA(
+		NameDispenser& _nameDispenser,
+		set<YulString> const& _variablesToReplace
+	):
+		m_nameDispenser(_nameDispenser), m_variablesToReplace(_variablesToReplace)
+	{ }
+
+	void operator()(FunctionDefinition& _function) override;
+	void operator()(ForLoop& _forLoop) override;
+	void operator()(Switch& _switch) override;
+	void operator()(Block& _block) override;
+
+private:
+	NameDispenser& m_nameDispenser;
+	set<YulString> const& m_variablesToReplace;
+	/// Variables (that are to be replaced) currently in scope.
+	set<YulString> m_variablesInScope;
+	/// Set of variables that do not have a specific value.
+	set<YulString> m_variablesToReassign;
+};
+
+void IntroduceControlFlowSSA::operator()(FunctionDefinition& _function)
+{
+	set<YulString> varsInScope;
+	std::swap(varsInScope, m_variablesInScope);
+	set<YulString> toReassign;
+	std::swap(toReassign, m_variablesToReassign);
+
+	for (auto const& param: _function.parameters)
+		if (m_variablesToReplace.count(param.name))
+		{
+			m_variablesInScope.insert(param.name);
+			m_variablesToReassign.insert(param.name);
+		}
+
+	ASTModifier::operator()(_function);
+
+	m_variablesInScope = std::move(varsInScope);
+	m_variablesToReassign = std::move(toReassign);
+}
+
+void IntroduceControlFlowSSA::operator()(ForLoop& _for)
+{
+	(*this)(_for.pre);
+
+	Assignments assignments;
+	assignments(_for.body);
+	assignments(_for.post);
+
+
+	for (auto const& var: assignments.names())
+		if (m_variablesInScope.count(var))
+			m_variablesToReassign.insert(var);
+
+	(*this)(_for.body);
+	(*this)(_for.post);
+}
+
+void IntroduceControlFlowSSA::operator()(Switch& _switch)
+{
+	yulAssert(m_variablesToReassign.empty(), "");
```

Comment:
Why is this true here but not in the beginning of `For`?

Reply:
Because only `for` creates a backwards edge in the control-flow graph.